aten/src/ATen/cudnn/Handle.cpp

```diff
@@ -34,8 +35,67 @@ struct Handle {
 };
 
 std::mutex mutex;
-std::unordered_map<int, Handle> handles;
 
+// Handles are lazily created as different threads request them,
+// but are never destroyed until the end of the program.
+// This is the desired behavior: We want unused handles to stick around,
+// so they can be reused by other threads later.
+//
+// To prevent potential deadlocks, we explicitly choose not to cap the number
+// of handles that are created per device.
+// Example of danger: If we cap the max handles at 4, and 5 threads are sharing a device,
+// only 4 can make forward progress at any time. The other 4 will not release their
+// handles until they exit, so the fifth cannot make progress until then.  This is
+// not a problem...UNLESS all 5 threads attempt some sort of synchronization at an
+// intermediate point (ie, before any of them have exited).  We have no way to anticipate
+// or enforce that user threads will not attempt such intermediate synchronization.
+// The only way to ensure safety is to avoid imposing a cap on the number of handles.
+std::unordered_map<int, std::stack<Handle>> created_handles;
+std::unordered_map<int, std::stack<cudnnHandle_t>> available_handles;
+
+struct PoolWindow
```

Comment:
Maybe some comment like, "PoolWindow lazily caches handles for a thread, so in the common case handle access doesn't incur a mutex lock."